src/Interpreters/MonotonousOrderByVisitor.h

```diff
@@ -0,0 +1,174 @@
+        ASTPtr & identifier;
+        Monotonicity & monotonicity;
+        DataTypePtr data_type = nullptr;
+        bool first = true;
+    };
+
+    static void visit(const ASTPtr & ast, Data & data)
+    {
+        auto * ast_function = ast->as<ASTFunction>();
+        if (ast_function)
+        {
+            visit(*ast_function, data);
+        }
+    }
+
+
+    static void visit(ASTFunction & ast_function, Data & data)
+    {
+        if (!data.monotonicity.is_always_monotonic)
+            return;
+
+        auto arguments = ast_function.arguments;
+        if (arguments->children.size() != 1)
+        {
+            data.monotonicity.is_always_monotonic = false;
+            return;
+        }
+
+        if (data.first)
+        {
+            data.identifier = ast_function.arguments->children[0];
+            if (!data.identifier->as<ASTIdentifier>())
+            {
+                data.monotonicity.is_always_monotonic = false;
+                return;
+            }
+        }
+
+        auto * identifier_ptr = data.identifier->as<ASTIdentifier>();
+
+        if (AggregateFunctionFactory::instance().isAggregateFunctionName(ast_function.name))
+        {
+            data.monotonicity.is_always_monotonic = false;
+            return;
+        }
+
+        const auto & function = FunctionFactory::instance().tryGet(ast_function.name, data.context);
+        auto pos = IdentifierSemantic::getMembership(*identifier_ptr->as<ASTIdentifier>());
+        if (pos == NULL)
+        {
+            data.monotonicity.is_always_monotonic = false;
+            return;
+        }
+
+        auto data_type_and_name = data.tables[*pos].columns.tryGetByName(identifier_ptr->shortName());
+
+        if (!data_type_and_name)
+        {
+            data.monotonicity.is_always_monotonic = false;
+            return;
+        }
+
+        if (data.first)
+        {
+            data.data_type = data_type_and_name->type;
+            data.first = false;
+        }
+        auto name = data_type_and_name->name;
+
+        ColumnsWithTypeAndName args;
+        args.emplace_back(data.data_type, name);
+        auto function_base = function->build(args);
+
+        auto cur_data = data.data_type;
+        data.data_type = function_base->getReturnType();
+
+        if (!function_base->hasInformationAboutMonotonicity())
+        {
+            data.monotonicity.is_always_monotonic = false;
+            return;
+        }
+
+        bool is_positive = data.monotonicity.is_positive;
+        data.monotonicity = function_base->getMonotonicityForRange(*cur_data, Field(), Field());
+
+        if (!is_positive)
+            data.monotonicity.is_positive = !data.monotonicity.is_positive;
+    }
+
+    static bool needChildVisit(const ASTPtr &, const ASTPtr &)
+    {
+        return true;
+    }
+
+};
+
+using MonotonicityCheckVisitor = InDepthNodeVisitor<MonotonicityCheckMatcher, false>;
+
+/// Finds SELECT that can be optimized
+class MonotonousOrderByData
```

Comment:
You should not use visitor here. You already have casted `select_query` in place you call `optimizeMonotonousFunctionsInOrderBy()`, so you can use it instead of useless AST traversal. It should be just a function over `ASTSelectQuery`.